sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -640,6 +641,196 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
+                              override val result: Seq[(Seq[Seq[String]], Int)])
+      extends AggCheck(rows, result) with ArrayCheckNested
+
+    val tableName = "test_agg_arr_collated"
+    val tableNameLowercase = "test_agg_arr_collated_lowercase"
+
+    Seq(
+      // simple
+      AggCheckSimple(
+        rows = Seq("array('aaa')", "array('AAA')"),
+        result = Seq((Seq("aaa"), 2))
+      ),
+      AggCheckSimple(
+        rows = Seq("array('aaa', 'bbb')", "array('AAA', 'BBB')"),
+        result = Seq((Seq("aaa", "bbb"), 2))
+      ),
+      AggCheckSimple(
+        rows = Seq("array('aaa')", "array('bbb')", "array('AAA')", "array('BBB')"),
+        result = Seq((Seq("aaa"), 2), (Seq("bbb"), 2))
+      ),
+      // nested
+      AggCheckNested(
+        rows = Seq("array(array('aaa'))", "array(array('AAA'))"),
+        result = Seq((Seq(Seq("aaa")), 2))
+      ),
+      AggCheckNested(
+        rows = Seq("array(array('aaa'), array('bbb'))", "array(array('AAA'), array('bbb'))"),
+        result = Seq((Seq(Seq("aaa"), Seq("bbb")), 2))
+      ),
+      AggCheckNested(
+        rows = Seq(
+          "array(array('aaa', 'aaa'), array('bbb', 'ccc'))",
+          "array(array('aaa', 'aaa'), array('bbb', 'ccc'), array('ddd'))",
+          "array(array('AAA', 'AAA'), array('BBB', 'CCC'))"
+        ),
+        result = Seq(
+          (Seq(Seq("aaa", "aaa"), Seq("bbb", "ccc")), 2),
+          (Seq(Seq("aaa", "aaa"), Seq("bbb", "ccc"), Seq("ddd")), 1)
+        )
+      )
+    ).map((check: AggCheck) =>
+        withTable(tableName, tableNameLowercase) {
+          def checkResults(table: String): Unit = {
+            checkAnswer(sql(s"select a, count(*) from $table group by a"),
+              check.result.map{ case (agg, cnt) => Row(agg, cnt) })
+            checkAnswer(sql(s"select distinct a from $table"),
+              check.result.map{ case (agg, _) => Row(agg) })
+          }
+
+          // check against non-binary collation
+          sql(s"create table $tableName(a ${check.dataTypeCollated}) using parquet")
+          check.rows.map(row => sql(s"insert into $tableName(a) values($row)"))
+          checkResults(tableName)
+
+          // binary collation with values converted to lowercase should match the results as well
+          sql(s"create table $tableNameLowercase(a ${check.dataType}) using parquet")
+          check.rows.map(row =>
+            // scalastyle:off caselocale
+            sql(
+              s"""
+                 |insert into $tableNameLowercase(a)
+                 |values(${UTF8String.fromString(row).toLowerCase})
+                 |""".stripMargin)
+            // scalastyle:on caselocale
+          )
+          checkResults(tableNameLowercase)
+        }
+    )
+  }
+
+  test("Join on arrays of collated strings") {
+    abstract class JoinCheck(val leftRows: Seq[String],
+                             val rightRows: Seq[String],
+                             val resultRows: Seq[Any])
+      extends ArrayCheck
```

Comment:
```suggestion
    abstract class JoinCheck(
        val leftRows: Seq[String],
        val rightRows: Seq[String],
        val resultRows: Seq[Any])
      extends ArrayCheck
```
per https://github.com/databricks/scala-style-guide?tab=readme-ov-file#spacing-and-indentation and below too